### PR TITLE
ARROW-2240: [Python] Array initialization with leading numpy nan fails with exception

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by xhochy** [comment]
rebased

**Review by xhochy — APPROVED** [review]
+1, LGTM

will merge once the rebased build passes.

**Comment by wesm** [comment]
Going to take a look at this

**Comment by cpcloud** [comment]
looks like a linting error

**Comment by cpcloud** [comment]
fixing now

File: cpp/src/arrow/python/builtin_convert.cc [hunk 1]
```diff
@@ -88,7 +88,7 @@ class ScalarVisitor {
 
   Status Visit(PyObject* obj) {
     ++total_count_;
-    if (obj == Py_None) {
+    if (obj == Py_None || (PyFloat_Check(obj) && std::isnan(PyFloat_AsDouble(obj)))) {
```
Comment (wesm) [inline cpp/src/arrow/python/builtin_convert.cc]: This logic is slightly duplicated with `PyFloat_IsNaN`

File: cpp/src/arrow/python/python-test.cc [hunk 2]
```diff
@@ -367,5 +367,20 @@ TEST_F(DecimalTest, UpdateWithNaN) {
...
+  PyObject* list = list_ref.obj();
+  std::string str("str");
+
+  ASSERT_EQ(0, PyList_SetItem(list, 0, PyFloat_FromDouble(NAN)));
+  ASSERT_EQ(0, PyList_SetItem(list, 1, PyUnicode_FromString(str.c_str())));
+
+  std::shared_ptr<Array> out;
+  auto pool = default_memory_pool();
+  ASSERT_OK(ConvertPySequence(list, pool, &out));
+}
```
Comment (wesm) [inline cpp/src/arrow/python/python-test.cc]: Is this unit test redundant from the Python unit test below?

**Review by wesm — COMMENTED** [review]
Thanks for looking at this, a couple minor comments

File: cpp/src/arrow/python/python-test.cc (hunk 2 above)
Comment (cpcloud) [reply cpp/src/arrow/python/python-test.cc]: It is, I can remove the python ones. I personally find it easier to write the C API calls and debug those (if necessary) than to cross the language boundary

Comment (wesm) [reply cpp/src/arrow/python/python-test.cc]: OK, let's leave them both here then =)

File: cpp/src/arrow/python/builtin_convert.cc [hunk 3]
```diff
@@ -88,7 +88,7 @@ class ScalarVisitor {
 
   Status Visit(PyObject* obj) {
     ++total_count_;
-    if (obj == Py_None) {
+    if (obj == Py_None || internal::PyFloat_IsNaN(obj)) {
```
Comment (pitrou) [inline cpp/src/arrow/python/builtin_convert.cc]: Hmm... Ideally, we should handle all scalar types here, including decimal or np.float32?

**Comment by pitrou** [comment]
Is this a real bug at all? A NaN is still a floating-point "number", so it sounds reasonable to abort type inference when given both a FP number and a Python string...

As a point of comparison, it also fails when you pass an explicit type:
```
>>> pa.array([np.nan, 's'], type=pa.string())
Traceback (most recent call last):
  File "<ipython-input-18-015a2f2f7924>", line 1, in <module>
    pa.array([np.nan, 's'], type=pa.string())
  File "array.pxi", line 181, in pyarrow.lib.array
  File "array.pxi", line 36, in pyarrow.lib._sequence_to_array
  File "error.pxi", line 77, in pyarrow.lib.check_status
ArrowInvalid: /home/antoine/arrow/cpp/src/arrow/python/builtin_convert.cc:979 code: AppendPySequence(seq, size, real_type, builder.get())
/home/antoine/arrow/cpp/src/arrow/python/builtin_convert.cc:402 code: static_cast<Derived*>(this)->AppendSingle(ref.obj())
Non bytes/unicode value encountered: nan
```

**Comment by xhochy** [comment]
Although `np.nan` is actually a float value, it is used at various places to represent null entries in arrays that don't have a nullability mask.

**Comment by pitrou** [comment]
I suppose only in object arrays?
Currently `pa.array([np.nan])` and `pa.array([np.nan, 1])` both return a float64 array... Does the PR change this?

**Comment by wesm** [comment]
It's a bit of a slippery slope to be treating NaN unilaterally as null -- if we did that then `pa.array([nan])` should return a NullArray. It may make sense to have an option `nan_as_null` to toggle this behavior on and off (if off, then nan is treated as any other floating point value)

**Comment by xhochy** [comment]
+1 for the option. I guess with nulls we can never have a case that caters for everyone by default.

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**